src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java

```diff
@@ -31,10 +49,110 @@ public List<ScheduledCase> getCases(GetCasesRequest request) {
+            return CasesMapper.mapToCourtCase(savedHearingEntity, caseEntity);
+        } else {
+            return CasesMapper.mapToCourtCase(caseEntity);
+        }
+    }
+
+    @Transactional
+    @Override
+    public ScheduledCase addCaseOrUpdate(AddCaseRequest addCaseRequest) {
+        Optional<CaseEntity> existingCase = caseRepository.findByCaseNumberAndCourthouse_CourthouseName(
+            addCaseRequest.getCaseNumber(), addCaseRequest.getCourthouse());
+        if (existingCase.isPresent()) {
+            return updateCase(addCaseRequest, existingCase.get());
+        } else {
+            return addCase(addCaseRequest);
+        }
+    }
+
+    private ScheduledCase updateCase(AddCaseRequest addCaseRequest, CaseEntity existingCase) {
+        CaseEntity updatedCaseEntity = updateCaseEntity(addCaseRequest, existingCase);
+
+        if (StringUtils.isNotBlank(addCaseRequest.getCourtroom())) {
+            HearingEntity hearingEntity = updateOrCreateHearingEntity(addCaseRequest, updatedCaseEntity);
+            return CasesMapper.mapToCourtCase(hearingEntity, updatedCaseEntity);
+        }
+        return CasesMapper.mapToCourtCase(updatedCaseEntity);
+    }
+
+
+    private HearingEntity updateOrCreateHearingEntity(AddCaseRequest addCaseRequest, CaseEntity caseEntity) {
+
+        Optional<HearingEntity> hearing = hearingRepository.findAll().stream()
+            .filter(hearingEntity -> caseEntity.getId().equals(hearingEntity.getCourtCase().getId())
+                && hearingEntity.getHearingDate().equals(LocalDate.now())).findFirst();
+
+
+        return hearing.map(hearingEntity -> hearingRepository.saveAndFlush(mapToHearingEntity(
+            addCaseRequest,
+            caseEntity,
+            hearingEntity
+        ))).orElseGet(() -> saveNewHearingEntity(addCaseRequest, caseEntity));
+    }
+
+    private CaseEntity updateCaseEntity(AddCaseRequest addCaseRequest, CaseEntity existingCase) {
+        return caseRepository.saveAndFlush(mapAddCaseRequestToCaseEntity(addCaseRequest, existingCase));
+    }
+
+    private HearingEntity saveNewHearingEntity(AddCaseRequest addCaseRequest, CaseEntity caseEntity) {
+        return hearingRepository.saveAndFlush(mapToHearingEntity(addCaseRequest, caseEntity, new HearingEntity()));
+    }
+
+    private CaseEntity saveNewCaseEntity(AddCaseRequest addCaseRequest) {
+        CaseEntity caseEntity = mapAddCaseRequestToCaseEntity(addCaseRequest, new CaseEntity());
+        return caseRepository.saveAndFlush(caseEntity);
+    }
+
+    private HearingEntity mapToHearingEntity(AddCaseRequest addCaseRequest, CaseEntity caseEntity, HearingEntity hearingEntity) {
+        if (!StringUtils.isBlank(addCaseRequest.getCourtroom())) {
+            CourtroomEntity courtroomEntity = getExistingCourtroom(addCaseRequest);
+            if (courtroomEntity == null) {
+                throw new DartsApiException(CaseApiError.COURTROOM_PROVIDED_DOES_NOT_EXIST);
```

Comment:
I think we are meant to create the courtroom if it doesn't exist as we won't ever have a definitive list of them.

Reply:
ACCEPT